### PR TITLE
Add CI job with minimal dependencies

### LINKED ISSUE #2211: Add CI run with minimal dependencies
#2210 is triggered by not having dask installed. Our current CI setup can't test for this.

This should be addressed with a CI run that uses minimal dependencies.

### PR DESCRIPTION
**Pull request by flying-sheep**
Fixes #2211

I’ll first try with all removed. I expect the min_tests run to fail. Seeing what fails,

- If it’s not much, I’ll just refactor the fixtures a bit and so on
- Else I’ll move algorithms to the `test` extra. Then we can merge this PR and over time refactor our tests so more and more extras go from `tests` to `tests-full`

@ivirshup do you like the collection extras’ names (`io`, `speedups`, `algorithms`)? Should we add an extra named `all` that installs all the `-full` extras?

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -112,15 +110,37 @@ dev = [
...
+    "scanpy[skmisc]",
+    "scanpy[harmony]",
+    "scanpy[scanorama]",
+    "scanpy[scrublet]",
+]
+algorithms-full = [
+    "scanpy[algorithms]",
+    "scanpy[louvain]",
+    "scanpy[bbknn]",
+]
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
```

I think these are a bit much. I'd be up for `test` and `test-min`/ `test-full` or something along those lines.

It's just a lot of options that we're not going to maintain. Also I would like to remove a number of these.

File: pyproject.toml [hunk 2]
```diff
@@ -112,15 +110,37 @@ dev = [
...
 scrublet = ["scrublet"]  # Doublet detection
+# Acceleration
+speedups = ["scanpy[dask]", "scanpy[zappy]"]  # all CPU based speedups
+speedups-full = ["scanpy[speedups]", "scanpy[rapids]"]
+rapids = ["cudf>=0.9", "cuml>=0.9", "cugraph>=0.9"]  # GPU accelerated calculation of neighbors
+dask = ["dask[array]!=2.17.0"]  # Use the Dask parallelization engine
+zappy = ["zappy", "scanpy[zarr]"]  # Distributed array processing using Zarr
+# Optional I/O formats
+io = ["scanpy[zarr]"]  # all optional I/O formats
+zarr = ["zarr"]  # Support for the zarr data format
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
```

**Review by ivirshup — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (flying-sheep) [reply pyproject.toml]: I think it’s useful to enable people to install `scanpy[feature]` to get a feature.

And once we add the test extras, people will start to add `scanpy[test]` to get a “default feature set” which I think is ugly.

So I’ll do it basically as you want but a bit different.

File: pyproject.toml [hunk 3]
```diff
@@ -86,24 +84,36 @@ Home-page = "https://scanpy.org"
...
     "profimp",
-    # Test the metadata while this exists: https://github.com/takluyver/flit/issues/387
-    "flit_core",
+]
+test = [
+    "scanpy[test-min]",
+    # Optional but important dependencies
+    "scanpy[leiden]",
+    "zarr",
+]
```
Comment (flying-sheep) [inline pyproject.toml]: OK so what’s your plan with this?

- I just guessed the set of “optional but important” tests. Which ones should it be?
- Do we want to check if our set of “optional but important” tests aren’t skipped? If so, how?

File: pyproject.toml [hunk 4]
```diff
@@ -111,36 +123,21 @@ dev = [
...
-    "scanpy[louvain]",
-    "scanpy[bbknn]",
-]
+paga = ["python-igraph"]
 louvain = ["python-igraph", "louvain>=0.6,!=0.6.2"]  # Louvain community detection
 leiden = ["python-igraph", "leidenalg"]  # Leiden community detection
 bbknn = ["bbknn"]  # Batch balanced KNN (batch correction)
 magic = ["magic-impute>=2.0"]  # MAGIC imputation method
-skmisc = ["scikit-misc>=0.1.3"]  # highly_variable_genes method 'seurat_v3'
+seurat = ["scikit-misc>=0.1.3"]  # highly_variable_genes method 'seurat_v3'
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
skmisc = ["scikit-misc>=0.1.3"]  # highly_variable_genes method 'seurat_v3'
```

I feel like this name introduces more confusion than it solves.

File: pyproject.toml [hunk 5]
```diff
@@ -112,15 +122,22 @@ dev = [
...
 harmony = ["harmonypy"]  # Harmony dataset integration
 scanorama = ["scanorama"]  # Scanorama dataset integration
 scrublet = ["scrublet"]  # Doublet detection
+# Acceleration
+rapids = ["cudf>=0.9", "cuml>=0.9", "cugraph>=0.9"]  # GPU accelerated calculation of neighbors
+dask = ["dask[array]!=2.17.0"]  # Use the Dask parallelization engine
+zappy = ["zappy", "zarr"]  # Distributed array processing using Zarr
```
Comment (ivirshup) [inline pyproject.toml]: We shouldn't add a optional install for a deprecated library

```suggestion
```

File: pyproject.toml [hunk 6]
```diff
@@ -86,24 +84,36 @@ Home-page = "https://scanpy.org"
...
+    "zarr",
+]
+test-full = [
+    "scanpy[test]",
+    "scanpy[dask]",
+    "scanpy[zappy]",
+    # tested algorithms
+    "scanpy[leiden]",
+    "scanpy[magic]",
+    "scanpy[seurat]",
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
    "scanpy[skmisc]",
```

File: pyproject.toml [hunk 7]
```diff
@@ -86,24 +84,36 @@ Home-page = "https://scanpy.org"
...
+test = [
+    "scanpy[test-min]",
+    # Optional but important dependencies
+    "scanpy[leiden]",
+    "zarr",
+]
+test-full = [
+    "scanpy[test]",
+    "scanpy[dask]",
+    "scanpy[zappy]",
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 8]
```diff
@@ -112,15 +121,21 @@ dev = [
...
 magic = ["magic-impute>=2.0"]  # MAGIC imputation method
 skmisc = ["scikit-misc>=0.1.3"]  # highly_variable_genes method 'seurat_v3'
 harmony = ["harmonypy"]  # Harmony dataset integration
 scanorama = ["scanorama"]  # Scanorama dataset integration
 scrublet = ["scrublet"]  # Doublet detection
+# Acceleration
+rapids = ["cudf>=0.9", "cuml>=0.9", "cugraph>=0.9"]  # GPU accelerated calculation of neighbors
+dask = ["dask[array]!=2.17.0"]  # Use the Dask parallelization engine
+# Backwards compat
+skmisc = ["scanpy[seurat]"]
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
```

Missed one

**Review by ivirshup — APPROVED** [review]

File: pyproject.toml (hunk 3 above)
Comment (ivirshup) [reply pyproject.toml]: This looks fine for now.

I thought about it more and am not sure of the best approach so I'm good to merge this once the tests pass.

File: pyproject.toml (hunk 7 above)
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 9]
```diff
@@ -86,24 +84,36 @@ Home-page = "https://scanpy.org"
...
-    "zarr",
     "profimp",
-    # Test the metadata while this exists: https://github.com/takluyver/flit/issues/387
-    "flit_core",
+]
+test = [
+    "scanpy[test-min]",
+    # Optional but important dependencies
+    "scanpy[leiden]",
+    "zarr",
```
Comment (ivirshup) [inline pyproject.toml:96]:
```suggestion
    "zarr",
    "scanpy[dask]",
```

File: pyproject.toml [hunk 10]
```diff
@@ -86,24 +84,36 @@ Home-page = "https://scanpy.org"
...
+]
+test-full = [
+    "scanpy[test]",
+    "scanpy[dask]",
```
Comment (ivirshup) [inline pyproject.toml]:
```suggestion
```

**Comment by ivirshup** [comment]
@meeseeksdev backport to 1.9.x

**Pull request merged**